### PR TITLE
HRVO simulator support for World and PrimitiveSet

### LINKED ISSUE #2372: Remove raw pointers in HRVO implementation
### Description of the task
The HRVO implementation uses raw pointers. To avoid accidental memory leaks and have the code more maintainable long term, all pointers should be references or smart pointers.

### Acceptance criteria
- [ ] No `new` or `delete` calls
- [ ] Test the functionality to make sure that it remains unchanged

### LINKED ISSUE #2388: Add support for HRVO Agent moving at linear velocity
### Description of the task
Currently all robots (aka agents) in the HRVO simulator use the HRVO algorithm to avoid collision and go towards their `Goal` position, however, in real life, not all dynamic objects will be trying to avoid other obstacles. For example, we do not know the behaviour of the enemy robots, and as a result, predicting their movement using HRVO is likely to be inaccurate, thus we will be assuming that the enemy robots will be moving at a constant velocity. This will also be useful for avoiding the ball when it is an obstacle, since the ball will not be trying to dodge our robots.
Could also add constant acceleration/deceleration for the new Agent type to make it more realistic.

### Acceptance criteria
- [ ] Define a new method for defining HRVO Agents which will be moving at a linear velocity with constant acceleration
- [ ] Add tests

### PR DESCRIPTION
**Pull request by nimazareian**
### Description
Added support for updating the state of the HRVO simulator using `World` and `PrimitiveSet`.

Added support for linear velocity agent to the HRVO simulator. Linear velocity agents ignore all obstacles and only try to get to their destination. They are used to represent the enemy robots, since we can not guarantee that enemy robots will be trying to avoid collision with our robots. Further, it is assumed that the enemy robot will continue to move in its current direction until it reached the edge of the field, or the friendly defence area. Linear velocity agents are also used to represent the ball as an obstacle. Since linear velocity agents do not perform collision avoidance, HRVO agents use velocity obstacles to avoid colliding with them, instead of using hybrid reciprocal velocity obstacles.

The kinematics for calculating the preferred velocity for HRVO agents have also been improved (mainly during acceleration), which reduces collisions.

All usage of dynamic memory allocation has been replaced with smart pointers (`unique_ptr` specifically).

General code clean up and documentation.

### Testing Done
Did testing using a new set of visualization tests in `hrvo_test.cpp`.

You can run the visualization tests by 

1. first running `hrvo_test.cpp`: `bazel run //extlibs/hrvo:hrvo_test`
2. then running a desired test by running the following command from the `src` directory: `python ./extlibs/hrvo/visualizer/hrvo_visualizer.py -p="/tmp/_test_name_.csv"` (replacing _test_name_)
- eg. `python ./extlibs/hrvo/visualizer/hrvo_visualizer.py -p="/tmp/twenty_five_robots_around_circle.csv"`

### Resolved Issues
resolves #2388 
resolves #2372 

### Length Justification
`Agent` class was refactored, and most of its implementation was moved to `HRVOAgent`, so a large portion of the changes are refactoring.
Another large portion includes either new or rewritten visualization tests.

### Review Checklist
`simulator.h/cpp` has the main new logic for using World and PrimitiveSet.

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [X] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [X] **Remove all commented out code**
- [X] **Remove extra print statements**: for example, those just used for testing
- [X] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/extlibs/hrvo/agent.h [hunk 1]
```diff
@@ -1,78 +1,41 @@
...
+     * Constructor
+     *
+     * @param simulator          The simulator which this agent is a part of
+     * @param position           The starting position of this agent.
+     * @param radius             The radius of this agent.
+     * @param velocity           The initial velocity of this agent.
+     * @param prefVelocity       The preferred velocity of this agent.
+     * @param maxSpeed           The maximum speed of this agent.
+     * @param maxAccel           The maximum acceleration of this agent.
+     * @param goalIndex             The goal number of this agent.
```
Comment (itsarune) [inline src/extlibs/hrvo/agent.h]: could you clarify what a goalIndex is?

File: src/extlibs/hrvo/hrvo_agent.h [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        Vector2 position_;
+
+        // The number of the first velocity obstacle.
+        int velocityObstacle1_;
+
+        // The number of the second velocity obstacle.
+        int velocityObstacle2_;
+    };
+
+   public:  // A
```
Comment (itsarune) [inline src/extlibs/hrvo/hrvo_agent.h]: "//A"?

File: src/extlibs/hrvo/linear_velocity_agent.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    if (abs(pref_velocity_) > max_speed_)
+    {
+        pref_velocity_ = normalize(pref_velocity_) * max_speed_;
+    }
+
+    const float dv = abs(pref_velocity_ - velocity_);
+    if (dv != 0.f)
+    {
+        new_velocity_ = (1.0f - (max_accel_ * simulator_->timeStep_ / dv)) * velocity_ +
+                        (max_accel_ * simulator_->timeStep_ / dv) * pref_velocity_;
```
Comment (itsarune) [inline src/extlibs/hrvo/linear_velocity_agent.cpp]: could you comment some details about the calculations happening in this function

File: src/extlibs/hrvo/vector2.h [hunk 4]
```diff
@@ -295,6 +295,19 @@ inline float det(const Vector2 &vector1, const Vector2 &vector2)
...
 }
 
+/**
+ * Computes the dot product of two vectors
+ *
+ * @param vector_left  The left hand side vector of dot product
+ * @param vector_right  The right hand side vector of dot product
+ * @return  The dot product of the two vectors.
+ */
+inline float dotProduct(const Vector2 &vector_left, const Vector2 &vector_right)
```
Comment (jonathanlew) [inline src/extlibs/hrvo/vector2.h:306]: can we add a TODO to link to #2371

File: src/extlibs/hrvo/simulator.h [hunk 5]
```diff
@@ -275,158 +245,57 @@ class Simulator
...
+    // This scale is used to avoid close encounters, and reduce chance of collision
+    const float friendly_robot_radius_scale = 1.25f;
 
-    //	private:
-   public:
-    Simulator(const Simulator &other);
-    Simulator &operator=(const Simulator &other);
+    // The scale which enemy robots should be larger than their actual size
+    // This scale is used to avoid close encounters, and reduce chance of collision
+    const float enemy_robot_radius_scale = 2.f;
```
Comment (itsarune) [inline src/extlibs/hrvo/simulator.h]: can we get this from RobotNavigationObstacleConfig

**Review by itsarune — COMMENTED** [review]
nice work integrating it with Thunderbots code

File: src/extlibs/hrvo/linear_velocity_agent.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     * @param maxSpeed    The maximum speed of this agent.
+     * @param maxAccel    The maximum acceleration of this agent.
+     * @param goal_index  The goal number of this agent.
+     * @param goalRadius  The goal radius of this agent.
+     */
+    LinearVelocityAgent(Simulator *simulator, const Vector2 &position, float radius,
+                        const Vector2 &velocity, float maxSpeed, float maxAccel,
+                        std::size_t goal_index, float goalRadius);
+
+    virtual ~LinearVelocityAgent() override = default;
```
Comment (jonathanlew) [inline src/extlibs/hrvo/linear_velocity_agent.h]: we typically don't do this for our subclasses, so I'm not sure why it's needed

File: src/extlibs/hrvo/simulator.cpp [hunk 7]
```diff
@@ -36,96 +36,243 @@
...
+    agents_.push_back(std::move(agent));
     return agents_.size() - 1;
 }
 
 std::size_t Simulator::addGoal(const Vector2 &position)
 {
-    Goal *const goal = new Goal(position);
-    goals_.push_back(goal);
+    std::unique_ptr<Goal> goal = std::make_unique<Goal>(position);
+    goals_.push_back(std::move(goal));
```
Comment (jonathanlew) [inline src/extlibs/hrvo/simulator.cpp:248]: I'm confused by this, shouldn't goals be on a per agent basis?

**Review by jonathanlew — COMMENTED** [review]
just a few comments

File: src/extlibs/hrvo/hrvo_test.cpp [hunk 8]
```diff
@@ -173,86 +211,192 @@ class HRVOTest : public ::testing::Test
...
-    }
-    add_static_obstacle(Vector2(0, 2.f), 0.75f);
-    create_div_b_field();
+    std::vector<std::pair<Point, Point>> friendly_start_dest_points = {
+        std::pair(Point(0.0, 0.0), Point(0.0, 0.0)),
+        std::pair(Point(-2.0, 0.0), Point(4.0, 0.0))};
+    instantiate_robots_in_world(friendly_start_dest_points, {});
+}
+
+TEST_F(HRVOTest, stationary_friendly_robot_dodging_moving_enemy_robot)
```
Comment (jonathanlew) [inline src/extlibs/hrvo/hrvo_test.cpp:249]: It looks like the robots collide, when will there be validation for these tests?

File: src/extlibs/hrvo/hrvo_test.cpp [hunk 9]
```diff
@@ -173,86 +211,192 @@ class HRVOTest : public ::testing::Test
...
+}
+
+TEST_F(HRVOTest, single_friendly_robot_moving_in_line)
+{
+    std::vector<std::pair<Point, Point>> friendly_start_dest_points = {
+        std::pair(Point(-5.0, 0.0), Point(5.0, 0.0))};
+    instantiate_robots_in_world(friendly_start_dest_points, {});
+}
+
+TEST_F(HRVOTest, destination_between_friendly_robot_and_stationary_enemy_robot)
```
Comment (jonathanlew) [inline src/extlibs/hrvo/hrvo_test.cpp:289]: can we plot the goal position too?

File: src/extlibs/hrvo/hrvo_test.cpp [hunk 10]
```diff
@@ -173,86 +211,192 @@ class HRVOTest : public ::testing::Test
...
+
+    std::vector<std::pair<Point, Point>> friendly_start_dest_points = {
+        std::pair(Point(-5.0, 0.0), Point(4.0, 0.0))};
+    std::vector<std::pair<Point, Vector>> enemy_position_velocity_pairs = {
+        std::pair(Point(5.0, 0.0), Vector(0.0, 0.0))};
+    instantiate_robots_in_world(friendly_start_dest_points,
+                                enemy_position_velocity_pairs);
+}
+
+TEST_F(HRVOTest, destination_between_friendly_robot_and_moving_enemy_robot)
```
Comment (jonathanlew) [inline src/extlibs/hrvo/hrvo_test.cpp:302]: It looks like the robots collide

File: src/extlibs/hrvo/hrvo_test.cpp [hunk 11]
```diff
@@ -173,86 +211,192 @@ class HRVOTest : public ::testing::Test
...
+        enemy_position_velocity_pairs.emplace_back(
+            std::make_pair(enemy_pos, Vector(1.5, 0.0)));
+    }
+    instantiate_robots_in_world(friendly_start_dest_points,
+                                enemy_position_velocity_pairs);
+    setSimulationTimeout(7.f);
 }
 
-TEST_F(HRVOTest, 1_robot_moving_in_square_with_final_speed)
+TEST_F(HRVOTest, friendly_robot_going_around_ball_obstacle)
```
Comment (jonathanlew) [inline src/extlibs/hrvo/hrvo_test.cpp:407]: why is the ball the same size as the robot?

File: src/extlibs/hrvo/hrvo_test.cpp [hunk 12]
```diff
@@ -173,86 +211,192 @@ class HRVOTest : public ::testing::Test
...
-        } while (!simulator.haveReachedGoals() && prev_frame_time < 15.f);
+        } while (!simulator.haveReachedGoals() && prev_frame_time < simulation_timeout);
 
         auto finish_time = std::chrono::high_resolution_clock::now();
         std::chrono::duration<double> total_time = finish_time - start_time;
         std::cout << "Total run time = " << total_time.count() << std::endl;
         std::cout << "Total simulation time = " << prev_frame_time << std::endl;
-        std::cout << "Average time per tick = " << total_time.count() / frame
-                  << std::endl;
+        std::cout << "Avg time per tick = " << total_time.count() / frame << std::endl;
```
Comment (jonathanlew) [inline src/extlibs/hrvo/hrvo_test.cpp]: can we give units for all this time?

**Review by jonathanlew — COMMENTED** [review]
I just did a pass over the tests, they're really cool. Just some more comments

File: src/extlibs/hrvo/hrvo_test.cpp (hunk 8 above)
Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_test.cpp:249]: These tests serve as an easy way to visualize various scenarios, such as edge cases, so I did not add any validation. It would probably be better to have actual validation once HRVO is running on the ERForce Sim with the correct field and robot configuration, rather than the simplified HRVO simulator. That way we can also tune any values/formulas with a more realistic simulator.
About the robots colliding, I think the expected behavior would be that friendly robots do not dodge enemy robots once they have reached their destination. Though I haven't thought about it too much, so it's open for debate.

File: src/extlibs/hrvo/hrvo_test.cpp (hunk 10 above)
Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_test.cpp:302]: Replied about this scenario, above.

File: src/extlibs/hrvo/hrvo_test.cpp (hunk 11 above)
Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_test.cpp:407]: I ended up setting all Agent radii to be the same size as the max robot radius, since the actual Agent radius is generally scaled (eg. doubled for enemy agents, or 0.5m for ball). So the visualization would've been what HRVO sees, instead of the actual robot/ball size. To make it more realistic, I made a generalization that in the visualization, all Agents are the same size as the max robot radius.

File: src/extlibs/hrvo/simulator.cpp (hunk 7 above)
Comment (nimazareian) [reply src/extlibs/hrvo/simulator.cpp:248]: Yes it should. The original HRVO implementation had it implemented this way, where Agent and Goal are both owned by the Simulator, rather than each Agent owning its own Goal. This is being fixed by @JoshuaWu7 in #2373

File: src/extlibs/hrvo/vector2.h (hunk 4 above)
Comment (nimazareian) [reply src/extlibs/hrvo/vector2.h:306]: Added a TODO to the top of the file that the entire file should be removed with that ticket.

File: src/extlibs/hrvo/linear_velocity_agent.h (hunk 6 above)
Comment (nimazareian) [reply src/extlibs/hrvo/linear_velocity_agent.h]: Removed the destructors in the subclasses

File: src/extlibs/hrvo/hrvo_test.cpp (hunk 12 above)
Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_test.cpp]: Added units to the prints

File: src/extlibs/hrvo/hrvo_agent.h (hunk 2 above)
Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_agent.h]: Comment removed.

File: src/extlibs/hrvo/agent.h (hunk 1 above)
Comment (nimazareian) [reply src/extlibs/hrvo/agent.h]: Comment updated to be more descriptive

File: src/extlibs/hrvo/linear_velocity_agent.cpp (hunk 3 above)
Comment (nimazareian) [reply src/extlibs/hrvo/linear_velocity_agent.cpp]: I simplified the calculation here to make it more intuitive and added a comment.

File: src/extlibs/hrvo/simulator.h (hunk 5 above)
Comment (nimazareian) [reply src/extlibs/hrvo/simulator.h]: Honestly, I don't know how we're suppose to use the yaml configs. I'm also not sure what benefit it would give us to put the constants in a separate file. Would the goal be to be able to dynamically tune the values? 
For now I updated the constants to be `static constexpr`, similar to the constants in `theta_star_path_planner.h`.

File: src/extlibs/hrvo/hrvo_test.cpp (hunk 9 above)
Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_test.cpp:289]: I can look into it.

Comment (nimazareian) [reply src/extlibs/hrvo/hrvo_test.cpp:289]: I've added it. This was a good idea, its nice to be able to visualize the goal position. Makes it easier to see the oscillation aswell.

**Review by jonathanlew — APPROVED** [review]
LGTM! :railway_track:

File: src/extlibs/hrvo/simulator.h (hunk 5 above)
Comment (itsarune) [reply src/extlibs/hrvo/simulator.h]: it'd be nice to be dynamic but this could be a later ticket

**Review by itsarune — APPROVED** [review]
🙌🙌
